### PR TITLE
[core] prestart worker on node startup

### PR DESCRIPTION
**Pull request by clarng**
## Why are these changes needed?

Always prestart num_cpu of workers when raylet starts up. Previously we only start python workers on driver registration, or another worker submits a new task to this raylet. This has caused cold start issues described in https://github.com/ray-project/ray/issues/26262.

As part of this change, also did some needed cleanup to simplify the code / make this work
- removed start_initial_python_workers_for_first_job from ray.init(...)
- this is causing prestart to not work, since start_initial_python_workers_for_first_job defaults to false and is defaulted to true by ray client if there is no runtime env
-- there is no behavior change in this PR to how worker prestart interacts with runtime env
-- this doesn't seem to be big of a change to warrant api review : if customer is setting this in ray client, they should remove it
-- if someone wants to turn off worker prestart, they can do so by setting RAY_enable_worker_prestart to false

Benchmark : on prestarted raylet, measure time to start driver and start num_cpu tasks

Master: 2.09 sec
PR: 1.18 sec(56% of original startup time)

We don't measure other cases due to restrictions with today's working re-use due to worker cache key - something that needs to be addressed in follow up

## Related issue number

#26262
https://github.com/ray-project/ray/pull/31848

## Checks

- [x] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by clarng** [comment]
(still need to fix a few test failures)

**Comment by iycheng** [comment]
I think the API changes looks OK since it's only for Ray Client. I'm not sure how many users are using it now.

If we have concern about this, maybe another way is to keep this param and if it's set, throw exception and the users will know this change and just giving them enough knowledge about how to change the code.

File: src/ray/raylet/node_manager.cc
Comment (iycheng) [inline src/ray/raylet/node_manager.cc:427]: nit: just some concerns here. Now we need to call worker-pool start to start the worker. If we fail to call it, some function seems not working. (idle killing).
not big deal for now.

**Comment by iycheng** [comment]
Others LGTM. Clean change.

**Comment by rkooo567** [comment]
Actually API change is not great because it is a direct feature request from the ray client user (I remember I helped fixing this issue). We should not change our behavior secretly. 

Let me ping the user directly about this change and hear his thoughts.

File: python/ray/tests/test_client_builder.py
Comment (rkooo567) [inline python/ray/tests/test_client_builder.py:427]: instead of removing it, test it with prestart=False?

File: src/ray/raylet/worker_pool.cc
Comment (rkooo567) [inline src/ray/raylet/worker_pool.cc:791]: do we have unit test for this logic in worker_pool_test?

Comment (clarng) [reply src/ray/raylet/worker_pool.cc:791]: Seems unrelated to this PR

**Comment by clarng** [comment]
> Actually, we should communicate this change to the user who brought this problem before. It was a direct feature request from the ray client user (I remember I helped fixing this issue). We should not change our behavior secretly.
> 
> I think asking to set prestart=False is reasonable, but we should make sure to update the doc. (it should be done within this PR). And maybe we should communicate this change to the OSS slack? (seems to be a big change to me. For example, when users start a new cluster, they will see the memory usage is drastically increased across the cluster)
> 
> Let me ping the user directly about this change and hear his thoughts.

What is the problem? Are they complaining that we prestart workers?

**Comment by rkooo567** [comment]
yeah. I pinged the person (also cc'ed you), so we can hear from him directly and find the proper solution! I think he complained that workers are started even before we submit any job.

**Comment by rkooo567** [comment]
Btw the PR itself LGTM. So I will approve it after talking to the user

**Comment by cadedaniel** [comment]
> yeah. I pinged the person (also cc'ed you), so we can hear from him directly and find the proper solution! I think he complained that workers are started even before we submit any job.

Can we put this behind an env var? That way we can ask them to try out this + https://github.com/ray-project/ray/pull/33572

**Comment by clarng** [comment]
> 



> * enable_worker_prestart

Hmm I think we shouldn't block this PR on waiting to hear back from user

- they won't get this change unless they upgrade
- even if they upgrade immediately it will take weeks for the release to be ready, and if we block this PR now it won't go out till next next release anyway and that will be months away
- until then, we have weeks to inform that user 
- also, this change improves the experience of many users we should optimize for the benefit of the majority

**Comment by clarng** [comment]
> > yeah. I pinged the person (also cc'ed you), so we can hear from him directly and find the proper solution! I think he complained that workers are started even before we submit any job.
> 
> Can we put this behind an env var? That way we can ask them to try out this + #33572

Why is RAY_enable_worker_prestart not sufficient?

**Comment by clarng** [comment]
> yeah. I pinged the person (also cc'ed you), so we can hear from him directly and find the proper solution! I think he complained that workers are started even before we submit any job.

Hmm which channel did you cc'ed ? I don't think i got it

**Comment by clarng** [comment]
> Actually, we should communicate this change to the user who brought this problem before. It was a direct feature request from the ray client user (I remember I helped fixing this issue). We should not change our behavior secretly.
> 
> I think asking to set prestart=False is reasonable, but we should make sure to update the doc. (it should be done within this PR). And maybe we should communicate this change to the OSS slack? (seems to be a big change to me. For example, when users start a new cluster, they will see the memory usage is drastically increased across the cluster)
> 
> Let me ping the user directly about this change and hear his thoughts.

Chatted with @rkooo567 offline, will follow up docs change

**Comment by clarng** [comment]
@rkooo567 are we still blocked on waiting to hear back from user? Seems we haven't contacted them in 2 months. Is that necessary for this PR to merge?

@scv119 @cadedaniel

File: src/ray/raylet/worker_pool.cc
Comment (scv119) [inline src/ray/raylet/worker_pool.cc:164]: is it too much ask to add another flag controls when prestart happens (raylet or first driver registers), and default to first driver registers?

**Review by rkooo567 — APPROVED** [review]

**Comment by rkooo567** [comment]
let's make sure to handle https://github.com/ray-project/ray/pull/33623#discussion_r1148849057?

**Comment by clarng** [comment]
@rkooo567 @scv119  can we please merge this now?

**Comment by rkooo567** [comment]
Lmk when tests pass. I retried https://buildkite.com/ray-project/oss-ci-build-pr/builds/16563#01872646-3afa-4fd3-a47f-9c4c5c99ad3a windows build because it failed for weird reasons.

Also, I believe there's a better approach that works in general cases, but we can discuss it as a follow up after merging the PR. I will schedule a meeting to discuss.

**Comment by rkooo567** [comment]
several test failure seems related (e.g., worker_pool_test)

**Comment by clarng** [comment]
> several test failure seems related (e.g., worker_pool_test)

looks passing : https://github.com/ray-project/ray/pull/33623#issuecomment-1486221016

**Comment by rkooo567** [comment]
waiting for other tests to finish...

**Comment by clarng** [comment]
@rkooo567  I think it is good to merge now

**Pull request merged**
